addons/survey/models/survey_survey.py

```diff
@@ -1270,3 +1260,33 @@ def _handle_certification_badges(self, vals):
             # delete all challenges and goals because not needed anymore (challenge lines are deleted in cascade)
             challenges_to_delete.unlink()
             goals_to_delete.unlink()
+
+    # ------------------------------------------------------------
+    # TOOLING / MISC
+    # ------------------------------------------------------------
+
+    def _generate_session_codes(self, code_count=1, excluded_codes=False):
+        """ Generate {code_count} session codes for surveys.
+
+        We try to generate 4 digits code first and see if we have {code_count} unique ones.
+        Then we raise up to 5 digits if we need more, etc until up to 10 digits. """
+
+        self.flush_model(['session_code'])
+
+        session_codes = set()
+        excluded_codes = excluded_codes or set()
+        existing_codes = self.sudo().search_read(
+            [('session_code', '!=', False)],
+            ['session_code']
+        )
+        unavailable_codes = excluded_codes | {existing_code['session_code'] for existing_code in existing_codes}
+        for digits_count in range(4, 10):
+            range_lower_bound = 10 ** (digits_count - 1)
+            range_upper_bound = (range_lower_bound * 10) - 1
+            code_candidates = {str(random.randint(range_lower_bound, range_upper_bound)) for _ in range(20)}
```

Comment:
Why generating 20 candidates ?

Reply:
Good catch! That was actually a relic of the previous implementation where we generated codes one at a time and did not make sense anymore.

I modified it to "code_count + 20", to try to mitigate bad luck collisions in a best effort to keep shorter codes.
(There is probably statistically better than a fixed 20, likely a number scaly with the code_count, but heh, not sure we need to math it out that hard).

Thanks for the review!